java/vector/src/main/java/org/apache/arrow/vector/file/ArrowWriter.java

```diff
@@ -72,53 +69,26 @@ protected ArrowWriter(VectorSchemaRoot root, DictionaryProvider provider, Writab
     this.out = new WriteChannel(out);
 
     List<Field> fields = new ArrayList<>(root.getSchema().getFields().size());
-    Map<Long, ArrowDictionaryBatch> dictionaryBatches = new HashMap<>();
+    Set<Long> dictionaryIdsUsed = new HashSet<>();
 
+    // Convert fields with dictionaries to have dictionary type
     for (Field field: root.getSchema().getFields()) {
-      fields.add(toMessageFormat(field, provider, dictionaryBatches));
+      fields.add(DictionaryUtility.toMessageFormat(field, provider, dictionaryIdsUsed));
     }
 
-    this.schema = new Schema(fields, root.getSchema().getCustomMetadata());
-    this.dictionaries = Collections.unmodifiableList(new ArrayList<>(dictionaryBatches.values()));
-  }
-
-  // in the message format, fields have the dictionary type
-  // in the memory format, they have the index type
-  private Field toMessageFormat(Field field, DictionaryProvider provider, Map<Long, ArrowDictionaryBatch> batches) {
-    DictionaryEncoding encoding = field.getDictionary();
-    List<Field> children = field.getChildren();
-
-    if (encoding == null && children.isEmpty()) {
-      return field;
-    }
-
-    List<Field> updatedChildren = new ArrayList<>(children.size());
-    for (Field child: children) {
-      updatedChildren.add(toMessageFormat(child, provider, batches));
-    }
-
-    ArrowType type;
-    if (encoding == null) {
-      type = field.getType();
-    } else {
-      long id = encoding.getId();
+    // Create a record batch for each dictionary
+    this.dictionaries = new ArrayList<>(dictionaryIdsUsed.size());
+    for (long id: dictionaryIdsUsed) {
       Dictionary dictionary = provider.lookup(id);
-      if (dictionary == null) {
-        throw new IllegalArgumentException("Could not find dictionary with ID " + id);
-      }
```

Comment:
I removed it because the check is already done [here](https://github.com/apache/arrow/pull/752/files/8a323cf1d190a2133b620b980af2aaeebcff4ef8#diff-c47bcf8f4cc51762974cc2968da823f9R71) and the semantics are slightly different now:

1) Goes through each field in the schema
2) If field has an encoding, then verify the lookup dictionary in the provider (with the null check)
3)  Add the id to a set of dictionary ids that are being used by Fields
4) For each dictionary id being used, write the dictionary record batch (no need to check if this id is in the provider at this point because it was already done)

Hopefully that sounds ok